NPB/MG/mg.cpp

```diff
@@ -0,0 +1,1605 @@
+  }
+}
+
+/*--------------------------------------------------------------------
+c-------------------------------------------------------------------*/
+
+static void rprj3(dash::NArray<double, 3> &r, int m1k, int m2k, int m3k, dash::NArray<double, 3> &s, int m1j, int m2j, int m3j, int k) {
+  /*--------------------------------------------------------------------
+  c-------------------------------------------------------------------*/
+
+  /*--------------------------------------------------------------------
+  c   rprj3 projects onto the next coarser grid,
+  c   using a trilinear Finite Element projection:  s = r' = P r
+  c
+  c   This  implementation costs  20A + 4M per result, where
+  c   A and M denote the costs of Addition and Multiplication.
+  c   Note that this vectorizes, and is also fine for cache
+  c   based machines.
+  c-------------------------------------------------------------------*/
+
+  int d1, d2, d3;
+
+  if ( m1k == 3 ) {
+    d1 = 2;
+  } else {
+    d1 = 1;
+  }
+
+  if ( m2k == 3 ) {
+    d2 = 2;
+  } else {
+    d2 = 1;
+  }
+
+  if ( m3k == 3 ) {
+    d3 = 2;
+  } else {
+    d3 = 1;
+  }
+
+  int j2, j1, i3, i2, i1;
+  double x1[m1k], y1[m1k], x2, y2;
+
+  auto s_pattern = s.pattern();
+  auto s_local_beg_gidx = s_pattern.coords(s_pattern.global(0));
+  auto s_local_end_gidx = s_pattern.coords(s_pattern.global(s_pattern.local_size()-1));
+
+  auto r_pattern = r.pattern();
+  auto r_local_beg_gidx = r_pattern.coords(r_pattern.global(0));
+  auto r_local_end_gidx = r_pattern.coords(r_pattern.global(r_pattern.local_size()-1));
+  int r_offset = r_local_beg_gidx[0];
+  int r_x = (int) r.extent(2);
+  int r_y = (int) r.extent(1);
+
+  int start = 0;
+  if ( s_local_beg_gidx[0] == 0 ) start++;
+
+  int end = s.local.extent(0);
+  if ( s_local_end_gidx[0] == s.extent(0)-1 ) end--;
+
+  int r_planes = 2*(end-start)+1;
+
+  std::vector<std::vector<double> > r_local_v(r_planes);
+
+  for (int i = 0; i < r_planes; i++) {
+    r_local_v[i] = std::vector<double> (r_y*r_x);
+  }
+
+  std::vector<double*> r_local(r_planes);
+  std::vector<bool> is_local(r_planes);
+  int r_idx = 0;
+  int r_psize = r_y*r_x;
+
+  dash::Future<double*> futs[r_planes];
+
+  for (int j3l = start; j3l < end; j3l++){
+    int j3 = s_local_beg_gidx[0]+j3l;
+    i3 = 2*j3-d3;
+    if ( r(i3,0,0).is_local() ) {
+      r_local[r_idx] = r.lbegin()+r_psize*(i3-r_offset);
+      is_local[r_idx] = true;
+      futs[r_idx] = dash::copy_async(r.begin()+r_psize*i3, r.begin()+r_psize*i3, r_local_v[r_idx].data());
+    } else {
+      futs[r_idx] = dash::copy_async(r.begin()+r_psize*i3, r.begin()+r_psize*(i3+1), r_local_v[r_idx].data());
+      r_local[r_idx] = r_local_v[r_idx].data();
+    }
+    r_idx++;
+    if ( r(i3+1,0,0).is_local() ) {
+      r_local[r_idx] = r.lbegin()+r_psize*(i3+1-r_offset);
+      is_local[r_idx] = true;
+      futs[r_idx] = dash::copy_async(r.begin()+r_psize*(i3+1), r.begin()+r_psize*(i3+1), r_local_v[r_idx].data());
```

Comment:
```suggestion
      futs[r_idx] = dash::copy_async(r.begin()+r_psize*(i3+1), r.begin()+r_psize*(i3+2), r_local_v[r_idx].data());
```

Reply:
This is actually the work-around for setting the futures to true without having to do any calculations. Unfortunately, if we leave them out and just use the is_local[], it would give an even more ugly looking code when those elements are accessed. We couldn't find a better way to do it without working in DART or something similar.

Reply:
> This is actually the work-around for setting the futures to true without having to do any calculations.

I'm not sure I understand your point. Why do you need to a future in the first place then if there is no transfer happening? Such a quirk should at least be documented, preferably removed altogether.

Reply:
I looked it over and yes, you are right. I omitted the 0 element calls.